pkg/webhook/core.oam.dev/v1alpha2/traitdefinition/validating_handler.go

```diff
@@ -0,0 +1,219 @@
+type TraitDefValidator interface {
+	Validate(context.Context, v1alpha2.TraitDefinition) error
+}
+
+// TraitDefValidatorFn implements TraitDefValidator
+type TraitDefValidatorFn func(context.Context, v1alpha2.TraitDefinition) error
+
+// Validate implements TraitDefValidator method
+func (fn TraitDefValidatorFn) Validate(ctx context.Context, td v1alpha2.TraitDefinition) error {
+	return fn(ctx, td)
+}
+
+var _ admission.Handler = &ValidatingHandler{}
+
+// Handle validate trait definition
+func (h *ValidatingHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	obj := &v1alpha2.TraitDefinition{}
+	if req.Resource.String() != traitDefGVR.String() {
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("expect resource to be %s", traitDefGVR))
+	}
+	switch req.Operation { //nolint:exhaustive
+	case admissionv1beta1.Delete:
+		if len(req.OldObject.Raw) != 0 {
+			if err := h.Decoder.DecodeRaw(req.OldObject, obj); err != nil {
+				return admission.Errored(http.StatusBadRequest, err)
+			}
+		} else {
+			klog.Info("deleting Trait Definition", req.Name)
+		}
+	default:
+		err := h.Decoder.Decode(req, obj)
+		if err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+		for _, validator := range h.Validators {
+			if err := validator.Validate(ctx, *obj); err != nil {
+				klog.Info("validation failed ", " name: ", obj.Name, " errMsgi: ", err.Error())
+				return admission.Denied(err.Error())
+			}
+		}
+	}
+	return admission.ValidationResponse(true, "")
+}
+
+var _ inject.Client = &ValidatingHandler{}
+
+// InjectClient injects the client into the ValidatingHandler
+func (h *ValidatingHandler) InjectClient(c client.Client) error {
+	h.Client = c
+	return nil
+}
+
+var _ admission.DecoderInjector = &ValidatingHandler{}
+
+// InjectDecoder injects the decoder into the ValidatingHandler
+func (h *ValidatingHandler) InjectDecoder(d *admission.Decoder) error {
+	h.Decoder = d
+	return nil
+}
+
+// RegisterValidatingHandler will register TraitDefinition validation to webhook
+func RegisterValidatingHandler(mgr manager.Manager) error {
+	server := mgr.GetWebhookServer()
+	mapper, err := discoverymapper.New(mgr.GetConfig())
+	if err != nil {
+		return err
+	}
+	server.Register("/validating-core-oam-dev-v1alpha2-traitdefinitions", &webhook.Admission{Handler: &ValidatingHandler{
+		Mapper: mapper,
+		Validators: []TraitDefValidator{
+			TraitDefValidatorFn(ValidateDefinitionReference),
+			// add more validators here
+		},
+	}})
+	return nil
+}
+
+// ValidateDefinitionReference validates whether the trait definition is valid if
+// its `.spec.reference` field is unset.
+// It's valid if
+// it has at least one output, and all outputs must have GVK
+// or it has no output but has a patch
+// or it has a patch and outputs, and all outputs must have GVK
+func ValidateDefinitionReference(_ context.Context, td v1alpha2.TraitDefinition) error {
+	if td.Spec.Reference.Name == "" {
```

Comment:
```suggestion
	if td.Spec.Reference.Name != "" {
	  return nil
	}
```

We prefer to write clean code with less `{}`